src/extra/components/dld.py

```diff
@@ -0,0 +1,389 @@
+            raise ValueError('multiple detectors found, please pass one '
+                             'explicitly:\n' + ', '.join(sorted(detectors)))
+        elif detectors:
+            return detectors.pop()
+
+        raise ValueError('no detector found, please pass one explicitly')
+
+    def _align_pulse_index(self, kd, pulse_dim):
+        """Align pulse index to KeyData.
+
+        Args:
+            kd (extra_data.KeyData): KeyData object to align to.
+            pulse_dim ({pulseId, pulseIndex, time}, optional): Label
+                for pulse dimension, pulse ID by default.
+
+        Returns:
+            (pandas.Index or None): Index from internal pulses component
+                aligned to KeyData object in terms of trains.
+        """
+
+        if (data_counts := kd.data_counts()).sum() == 0:
+            return
+
+        pulse_counts = self._pulses.pulse_counts()
+
+        # Merge both counts in a dataframe joined over train ID.
+        entry_counts = pd.merge(
+            pulse_counts.rename('by_pulses'),
+            data_counts.rename('by_data'),
+            how='outer', left_index=True, right_index=True
+        ).fillna(-1).astype(int)
+
+        print(entry_counts)
+
+        # Only consider rows with data counts.
+        entry_counts = entry_counts[entry_counts['by_data'] > 0]
+
+        if (entry_counts['by_pulses'] == -1).any():
+            # Missing one or more trains in the pulse information.
+            raise ValueError('missing pulse information for one or more '
+                             'trains with data')
+
+        if len(entry_counts) < min(len(data_counts), len(pulse_counts)):
+            # Missing counts in one or more trains in data, select data
+            # and pulses down.
+            from extra_data import by_id
+            train_sel = by_id[entry_counts.index.to_numpy()]
+
+            pulses = self._pulses.select_trains(train_sel)
+            kd = kd.select_trains(train_sel)
+        else:
+            pulses = self._pulses
+
+        return pulses.build_pulse_index(pulse_dim)
+
+    def _build_reduced_pd(self, data, index, entry_level):
+        """Get variable-length data as series or dataframe.
+
+        Load and reduce data with a variable number of actual entries
+        among a fixed number of potential entries to build a linear
+        pandas Series or DataFrame.
+
+        The data must be 2-dimensional, with the first axis is assumed
+        to be an entry ("pulse") dimension and the second to only
+        contain actual data up to a variable number for each entry.
+
+        The data type must either be a scalar floating type (returning
+        Series objects) or a structured data type (returning DataFrame
+        objects) containing at least a single floating type.
+
+        Args:
+            key (str): Instrument source key.
+            pulse_dim (str): Pulse to use in index, must be supported by
+                underlying pulses object.
```

Comment:
These two parameters don't match the actual signature.

Reply:
Oops, forgot to update the docstring here.